### PR TITLE
Add optional other markdown editor choosen in project config.

### PR DESCRIPTION
**Pull request by olymk2**
#587 putting in this PR for comparison and discussion not ready to merge, needs testing nd style fixes at least and some unit tests.

### CODE REVIEW

**Comment by benjaoming** [comment]
Could you explain what Martor is and does?

**Comment by olymk2** [comment]
sure martor is just another markdown editor, it used to be called draceditor its what i actually use on my site I thought i used simplemde but must have moved to martor at some point.

Its has a really nice clean interface to it.

https://github.com/agusmakmun/django-markdown-editor

**Comment by benjaoming** [comment]
@olymk2 ah cool, so it's different from SimpleMDE? And the comparison and discussion you refer to is about which editor to use?

**Comment by olymk2** [comment]
yes its different, well the fact that django wiki allows pluggable editor means comparison / discussion over which on to use is not needed.

You can just choose from the ones provided, i am adding these to test using alternative editors and adjust things as needed.

Failing that this can be a starting point for some one else if i get stuck :)

**Comment by benjaoming** [comment]
@olymk2 cool, let's maybe share some criteria for a future default editor? What's your experience here?

I would ideally like something that...:

1. ...is code oriented - we chose Markdown specifically because it's simple and easy to read. This means that we can show code to normal users.
1. ...does not freak out about custom markdown tags like `[image id:n]`
1. ...provides a simple button interface for common tags (like this Github editor) and has keyboard shortcuts
1. ...has proven its stability and can be kept updated and is not expected to break in deployments a year down the road. MarkItUp is really nice because it hasn't broken a single time!
1. ...does not have a single button that doesn't work or is never needed
1. ...integrates with the remaining layout

**Comment by olymk2** [comment]
My experience is I have played around over the years mainly with markdown or org prefer org because its less fractured and has more useful feature but it a bit niche.

For me consistency is what i want, some of the users of the site are not that techy so require previews and button bars, so that the main thing i am after the current setup works but i keep getting asked how to change things because there is no button bar.

line numbers would also be nice, but mainly consistency across the site and ease of use, the default to me does not matter as long as there are options to change, having simplemde on part of my site then martor in another area does not look great.

I would also like to see consistency between create and edit as currently they behave quite different.

But my requirements are very similar to your list above.

**Comment by benjaoming** [comment]
Thanks for sharing your experience with Markdown! Yeah, the lack of an exhaustive button menu bar is a problem, agreed.

The numbered lines are interesting because:

1. They highlight the syntax of lines
1. Potentially they can be used for annotations (which requires another capability-like interface for the Editor class)

**Comment by olymk2** [comment]
Okay got the image bar working with martor now, so feel free to pull and run the test project switch between the three editor, no reason not to leave the default and let users decide.

martor still has at least a styling issue it button bar is quite long so goes under the sidebar, not actually that keen on the sidebar wonder if it could expand or something it takes up quite a bit of the screen width.

anyway be interested if anyone tries and finds any issues, probably need some test, what would we want to test ? this is frontend and it did not look like we do any selenium style tests, to test the the javascript / editor is integrated and working.

**Comment by olymk2** [comment]
I was going to write some selenium tests for my changes, anyone able to tell me how to run the tests ?

all i get is the error below, using make test py.test and any other method i try am i doing something wrong ?

`E   ImproperlyConfigured: Requested setting DEFAULT_INDEX_TABLESPACE, but settings are not configured. You must either define the environment variable DJANGO_SETTINGS_MODULE or call settings.configure() before accessing settings.`

using these instructions http://django-wiki.readthedocs.io/en/latest/development/testing.html

**Comment by benjaoming** [comment]
@olymk2 it's a problem mentioned in  #644 - you should run for instance `tox -e py35-django111` to test with Python 3.5 and Django 1.11

**Comment by olymk2** [comment]
that works unfortunately the selenium tests fail, although i have chromedriver installed and can fire up python an use it myself. guessing this is something todo with tox ?

any other way to run these tests with out tox ?


```
E       selenium.common.exceptions.WebDriverException: Message: unknown error: Chrome failed to start: exited abnormally
E         (Driver info: chromedriver=2.33.506092 (733a02544d189eeb751fe0d7ddca79a0ee28cce4),platform=Linux 4.13.0-16-lowlatency x86_64)

```

**Comment by benjaoming** [comment]
@olymk2 Is it possibly something missing in the tox environment that's created? Remember that it's created according to the requirements in `tox.ini `

**Comment by benjaoming** [comment]
@olymk2 you could try all this again with the current `master`. Stuff has been fixed so it's easier to run tests locally. See: http://django-wiki.readthedocs.io/en/master/development/index.html#get-started

File: src/wiki/editors/martor.py [hunk 1]
```diff
@@ -16,36 +16,32 @@ def get_admin_widget(self, instance=None):
...
-            'martor/js/martor.min.js',
-        )
+
+        if settings.EDITOR_INCLUDE_JAVASCRIPT is False:
+            m.Media.js = ()
         return m
 
     def get_urls(self):
-        return [url(r'^martor/', include('martor.urls')),]
+        return [url(r'^martor/', include('martor.urls')), ]
```
Comment (atombrella) [inline src/wiki/editors/martor.py]: It doesn't seem to be necessary?

**Comment by benjaoming** [comment]
Closing this because of the pro-longed staleness, but thanks for the work on this. The big issue seems to remain: To identify an API for plugins to add buttons and behaviors to these editors, either to the editors themselves, or at least have some kind of JS API where the plugins can add their contents at the current cursor position.

**Pull request closed without merging**